### PR TITLE
Add Azure DNS API

### PR DESCRIPTION
**Pull request by martgras**
Adding support for Azure DNS. 
Tested on debian, bash for Windows and pfsense.
Will also submit https://github.com/martgras/acme.sh/wiki/How-to-use-Azure-DNS

### CODE REVIEW

File: dnsapi/dns_azure.sh [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    return 1
+   fi
+   _debug _domain_id "$_domain_id"
+   _debug _sub_domain "$_sub_domain"
+   _debug _domain "$_domain"
+
+   acmeRecordURI="https://management.azure.com$(printf '%s' $_domain_id |sed 's/\\//g')/TXT/$_sub_domain?api-version=2017-09-01"
+   _debug $acmeRecordURI
+   body="{\"properties\": {\"TTL\": 3600, \"TXTRecords\": [{\"value\": [\"$txtvalue\"]}]}}"
+   _azure_rest PUT "$acmeRecordURI" "$body" "$accesstoken"
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: please check the return code  and the result in the response to make sure whether the PUT request succeed or not.

File: dnsapi/dns_azure.sh [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+   fi
+   _debug _domain_id "$_domain_id"
+   _debug _sub_domain "$_sub_domain"
+   _debug _domain "$_domain"
+
+   acmeRecordURI="https://management.azure.com$(printf '%s' $_domain_id |sed 's/\\//g')/TXT/$_sub_domain?api-version=2017-09-01"
+   _debug $acmeRecordURI
+   body="{\"properties\": {\"TTL\": 3600, \"TXTRecords\": [{\"value\": [\"$txtvalue\"]}]}}"
+   _azure_rest DELETE "$acmeRecordURI" "" "$accesstoken"
+   _debug $response
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: check the response too

File: dnsapi/dns_azure.sh [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+   export _H1="accept: application/json"
+   export _H2="Content-Type: application/x-www-form-urlencoded"
+   export _H3=""
+
+   body="resource=$(printf "%s" 'https://management.core.windows.net/'| _url_encode)&client_id=$(printf "%s" $clientID | _url_encode)&client_secret=$(printf "%s" $clientSecret| _url_encode)&grant_type=client_credentials"
+   _debug data "$body"
+   response="$(_post "$body" "https://login.windows.net/$TENANTID/oauth2/token" "" "POST" )"
+   accesstoken=$(printf "%s\n" "$response" | _egrep_o "\"access_token\":\"[^\"]*\"" | head -n 1 | cut -d : -f 2 | tr -d \")
+
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: check the response and check if the "accesstoken" is empty here.

File: dnsapi/dns_azure.sh (hunk 1 above)
Comment (martgras) [reply dnsapi/dns_azure.sh]: Actually I was looking for a way to get the http response code from _post or _get but didn't find a way 
Shouldn't this take care of handling the error 
Line 172 in _azure_rest() 

    if [ "$?" != "0" ]; then
      _err "error $ep"
    return 1

Other suggestions are welcome :)

File: dnsapi/dns_azure.sh [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+   _debug _domain "$_domain"
+
+   acmeRecordURI="https://management.azure.com$(printf '%s' $_domain_id |sed 's/\\//g')/TXT/$_sub_domain?api-version=2017-09-01"
+   _debug $acmeRecordURI
+   body="{\"properties\": {\"TTL\": 3600, \"TXTRecords\": [{\"value\": [\"$txtvalue\"]}]}}"
+   _azure_rest PUT "$acmeRecordURI" "$body" "$accesstoken"
+   if [ "$_code" = "200" ] || [ "$code" = '201' ]; then
+        _info "validation record added"
+   else
+        _err "error adding validation record ($_code)"
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: if  error happens,  please `return 1`

File: dnsapi/dns_azure.sh [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+## Ref: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-protocols-oauth-service-to-service#request-an-access-token
+_azure_getaccess_token() {
+   TENANTID=$1
+   clientID=$2
+   clientSecret=$3
+
+   export _H1="accept: application/json"
+   export _H2="Content-Type: application/x-www-form-urlencoded"
+   export _H3=""
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: why is there a empty `_H3` ?  can we remove it ?

File: dnsapi/dns_azure.sh [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+   body="resource=$(printf "%s" 'https://management.core.windows.net/'| _url_encode)&client_id=$(printf "%s" $clientID | _url_encode)&client_secret=$(printf "%s" $clientSecret| _url_encode)&grant_type=client_credentials"
+   _debug data "$body"
+   response="$(_post "$body" "https://login.windows.net/$TENANTID/oauth2/token" "" "POST" )"
+   accesstoken=$(printf "%s\n" "$response" | _egrep_o "\"access_token\":\"[^\"]*\"" | head -n 1 | cut -d : -f 2 | tr -d \")
+   _debug2 "response $response"
+   
+   if [ -Z "$accesstoken" ] ; then
```
Comment (Neilpang) [inline dnsapi/dns_azure.sh]: why is the `-Z` option  upper case ?

File: dnsapi/dns_azure.sh (hunk 4 above)
Comment (martgras) [reply dnsapi/dns_azure.sh]: thx - done

**Comment by Neilpang** [comment]
Please fix the travis  format errors:  https://travis-ci.org/Neilpang/acme.sh/jobs/332801384

And the add it into the dns api list in the main readme file.

Thanks.

**Comment by martgras** [comment]
travis is still showing failed but I can't find  why  - ( not familiar with travis)

**Pull request merged**